config/tests/src/test/java/org/jboss/windup/config/CategoriesTest.java

```diff
@@ -0,0 +1,282 @@
+    {
+        testProvider(TestCategory1RuleProvider.class, "Foo");
+    }
+
+    @Test
+    public void test2Category() throws IOException
+    {
+        testProvider(TestCategory2RuleProvider.class, "Foo, Bar");
+    }
+
+    @Test
+    public void testEnhanceUnsetCategory() throws IOException
+    {
+        testProvider(TestCategoryEnhanceUnsetProvider.class, "");
+    }
+
+    @Test
+    public void testEnhanceEmptyCategory() throws IOException
+    {
+        testProvider(TestCategoryEnhanceEmptyProvider.class, "");
+    }
+
+    @Test
+    public void testEnhanceStringCategory() throws IOException
+    {
+        testProvider(TestCategoryEnhanceStringRuleProvider.class, "Foo");
+    }
+
+    @Test
+    public void testEnhanceJoinedCategory() throws IOException
+    {
+        testProvider(TestCategoryEnhanceJoinedStringRuleProvider.class, "Foo, Bar");
+    }
+
+    @Test
+    public void testEnhanceCollectionCategory() throws IOException
+    {
+            testProvider(TestCategoryEnhanceCollectionRuleProvider.class, "Foo, Bar");
+    }
+
+
+
+    private void testProvider(final Class<? extends WindupRuleProvider> provider, final String expectedCatsStr) throws IOException
+    {
+        String name = TestCategoryUnsetRuleProvider.class.getSimpleName();
+
+        //final Set<String> expectedCats = new HashSet(Arrays.asList(expectedCatsStr.split("\\s*,\\s*+")));
+        final Set<String> expectedCats = new HashSet(Arrays.asList(StringUtils.split(expectedCatsStr, ", ")));
+
+        try (GraphContext context = factory.create())
+        {
+            Predicate<WindupRuleProvider> rulesToRun = new EnumeratedRuleProviderFilter(provider);
+
+            Configuration conf = loader.loadConfiguration(context, rulesToRun).getConfiguration();
+
+            Assert.assertTrue("Rule loaded", !conf.getRules().isEmpty());
+
+            // As wierd as it is, the object itself bears the behavior of a map, which is referred to as "context".
+            Assert.assertTrue("Rule instanceof Context", conf.getRules().get(0) instanceof Context);
+            Context subRule = (Context) conf.getRules().get(0);
+
+            Object category = subRule.get(RuleMetadata.CATEGORY);
+            Assert.assertTrue("CATEGORY is set in the Context: " + category, category != null);
+            Assert.assertTrue("Category is a Set: " + category.getClass(), category instanceof Set);
+            Set<String> cats = (Set<String>)category;
+            Assert.assertEquals("" + expectedCats.size() + " categories in " + name, expectedCats.size(), cats.size());
+            for (String expCat : expectedCats)
+                Assert.assertTrue(name + " has category " + expCat, cats.contains(expCat));
+        /*
+        /**/
+        }
+        catch(Exception ex) {
+            if(ex instanceof InvocationTargetException)
+                throw new WindupException("" + ((InvocationTargetException)ex).getTargetException(), ex);
```

Comment:
Because Forge/Furnace still gives "null" message and I have no idea what's not loaded.